packages/shell-api/src/aggregation-cursor.ts

```diff
@@ -42,7 +42,7 @@ export default class AggregationCursor extends ShellApiClass {
   }
 
 
-  async toReplString(): Promise<any> {
+  async asPrintable(): Promise<any> {
```

Comment:
The problem is that we want to override `asPrintable` per shell class, but we don't need to override `asShellResult` since that is generated in the decorator. I'd like to keep the decorators as fine-tuned as possible so we don't have to write a bunch of repeated code.  Right now `asPrintable` is a private internal method that is used to generate the `asShellResult` class, and is never called anywhere but the decorator. Does that still implicitly and magically coupled? It's only a single public method, `asShellResult`. 

An alternative would be to get rid of `asShellResult` entirely and just have `asPrintable`, then have the consumers (cli-repl/browser-repl) construct their own result via `{ type: evalResult.shellApiName, value: await evalResult.asPrintable() }`.

From a consumer perspective, I think having `asShellResult` is a bit nicer. Both ways will end up with the same definition of `asPrintable` in the ShellApi classes.

Reply:
Don't get me wrong I believe the way you choose is very good. My only concern there is how to make what's going on explicit and clear. It would be nice if who reads and implements the classes could do that without needing much context to do so correctly. But other than that i would not change anything. 

I would really avoid "cli-repl/browser-repl should not construct their own result via { type: evalResult.shellApiName, value: await evalResult.asPrintable() }.", not much so as it could be a design mistake, rather it could be a more concrete one: for example if the evaluation context would end up being totally isolated in some platforms it may end up being impossible to call methods on returned objects like `evalResult.asPrintable`.

Reply:
I think what will help _a lot_ with this is some documentation in `shell-api` and `shell-evaluator`. I think people can implement classes with ease if we are able to provide them with some instructions as to how to do it. (I'll make a ticket)

Reply:
Ok so I've reverted removing `asPrintable` because it's used directly in the java shell and I think it would be best to keep that the same. I've added comments wherever the asPrintable method is defined (i.e. all classes that need special printing) and also comments wherever asShellResult is defined (i.e. all classes that can't use the decorators, these are special cases). Going forward asShellResult should *only* be generated and not defined in the shell class.

This means that the consumers just call `asShellResult` as before. `asPrintable` is used only internally to define the `asShellResult` method and in the java shell. I think this is the cleanest because having each class define asShellResult seemed clunky, since the `asPrintable` method is pretty self explanatory but `asShellResult` is less so.